# A demonstration of ChEMBL webresource client

ChEMBL webresource client is the official Python cliet library that helps access ChEMBL data. It provides a simple interface to the ChEMBL web services and allows to query the ChEMBL database and retrieve data in a programmatic way. Adapted from the official [demo](http://beta.mybinder.org/v2/gh/chembl/chembl_webresource_client/master?filepath=demo_wrc.ipynb) included in the repo of [`chembl_webresource_client`](https://github.com/chembl/chembl_webresource_client), this jupyter notebook presents some more detailed examples of how to use the client library to access ChEMBL data.

## Available filters
ChEMBL webresource client provides a number of lookups for flexible quering of the database, making it filter and retrieve specific data. Here are some notes about each lookup type supported by ChEMBL:
  - `exact`: Matches an exact value, e.g., `name__exact='Aspirin'` would match if the name is exactly `'Aspirin'`.
  - `iexact`: Case-insensitive exact value, e.g., `name__iexact='aspirin'` would match if the name is `'aspirin'`, `'ASPIRIN'`, etc.
  - `contains`: Checks if a value contains the specified substring, e.g., `description__contains='pain'` would match if the description contains the substring `'pain'`.
  - `icontains`: Case-insensitive contains, e.g., `description__icontains='Pain'` would match if the description contains `'pain'`, `'Pain'`, etc.
  - `in`: Matches if the value is within a specified list, e.g., `id__in=[1, 2, 3]` would match if the id is 1, 2, or 3.
  - `gt`/`gte/lt/lte`: Greater than/greater than or equal to/less than/less than or equal to, e.g., `value__gt=10` would match if the value is greater than 10.
  - `startswith/endswith`: Matches if a value starts/ends with the specified substring, e.g., `name__startswith='Asp'` would match if the name starts with `'Asp'`.
  - `istartswith/iendswith`: Case-insensitive starts/ends with, e.g., `name__istartswith='asp'` would match if the name starts with `'asp'`, `'Asp'`, etc.
  - `range`: Matches if a value is within a specified range, e.g., `value__range=(1, 10)` would atch if the value is between 1 and 10 inclusive. 
  - `isnull`: Matches if a value is null, e.g., `date_isnull=True` would match if the date is null.
  - `regex`: Matches if a value matches the specified regular expression, e.g.,  `name__regex=r'^[A-Z]'` would match if the name starts with an uppercase letter.
  - `iregex`: Case-insensitive regular expression match, e.g., `name__iregex=r'^[a-z]'` would match if the name starts with a letter, regardless of case.
  - `search`: Full-text search, supported in som backends like PostgresSQL, e.g., `description__search='pain relief'` would match if the description matches the full-text search term `'pain relief'`.


## The `only` operator
### Introduction
The `only` method in the ChEMBL webresrouce client is used to limit the results of a query to a specific set of fields. The `only` method takes a single argument, which is a list of fields that one wants to include in the result. When one uses the `only` method to specify a subset of fields, the API will return only those fields, rather than the entire dataset. This reduces the amount of data being transmitted over the network, thus saving bandwidth and making the API call/query faster.

### Example

In [2]:
from chembl_webresource_client.new_client import new_client

# Create a client for molecules
molecule = new_client.molecule

mol_1 = molecule.get('CHEMBL25')
mol_2 = molecule.filter(chembl_id='CHEMBL25').only(['chembl_id', 'molecule_properties', 'molecule_structures'])
mol_3 = molecule.filter(chembl_id='CHEMBL25')

- As can be checked, `mol_1` is a dictionary containing 36 keys relevant to the compound aspirin, including `'atc_classifications'`, `'availability_type'`, ... `'molecule_properties'`, `'molecule_structures'`, ... etc.

In [3]:
mol_1.keys()

dict_keys(['atc_classifications', 'availability_type', 'biotherapeutic', 'black_box_warning', 'chebi_par_id', 'chemical_probe', 'chirality', 'cross_references', 'dosed_ingredient', 'first_approval', 'first_in_class', 'helm_notation', 'indication_class', 'inorganic_flag', 'max_phase', 'molecule_chembl_id', 'molecule_hierarchy', 'molecule_properties', 'molecule_structures', 'molecule_synonyms', 'molecule_type', 'natural_product', 'oral', 'orphan', 'parenteral', 'polymer_flag', 'pref_name', 'prodrug', 'structure_type', 'therapeutic_flag', 'topical', 'usan_stem', 'usan_stem_definition', 'usan_substem', 'usan_year', 'withdrawn_flag'])

- On the other hand, `mol_2` is an object of `chembl_webresource_client.query_set.QuerySet`. It can be indexed and has only one element in this case. Specifically, `mol_2[0]` is a dictionary containing only two keys (`'molecule_properties'` and `'molecule_structures'` for the same compound aspirin.)

In [4]:
type(mol_2)

chembl_webresource_client.query_set.QuerySet

In [5]:
mol_2[0]

{'molecule_properties': {'alogp': '1.31',
  'aromatic_rings': 1,
  'cx_logd': '-2.16',
  'cx_logp': '1.24',
  'cx_most_apka': '3.41',
  'cx_most_bpka': None,
  'full_molformula': 'C9H8O4',
  'full_mwt': '180.16',
  'hba': 3,
  'hba_lipinski': 4,
  'hbd': 1,
  'hbd_lipinski': 1,
  'heavy_atoms': 13,
  'molecular_species': 'ACID',
  'mw_freebase': '180.16',
  'mw_monoisotopic': '180.0423',
  'np_likeness_score': '0.12',
  'num_lipinski_ro5_violations': 0,
  'num_ro5_violations': 0,
  'psa': '63.60',
  'qed_weighted': '0.55',
  'ro3_pass': 'N',
  'rtb': 2},
 'molecule_structures': {'canonical_smiles': 'CC(=O)Oc1ccccc1C(=O)O',
  'molfile': '\n     RDKit          2D\n\n 13 13  0  0  0  0  0  0  0  0999 V2000\n   19.8052   -4.2758    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   19.8040   -5.0953    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   20.5121   -5.5043    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n   21.2217   -5.0948    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n  

- `mol_3` is also an object of `chembl_webresource_client.query_set.QuerySet` and also has only one element, which is a dictionary for the compound aspirin. However, since the `only` method was not used. `mol_3[0]` is equivalent to `mol_1`.

In [6]:
mol_3[0] == mol_1

True

- So for `mol_1` and `mol_3`, the query retrieved all information about the molecule with ChEMBL ID `CHEMBL25`, which can include a lot of data. On the other hand, for `mol_2`, the query was limited to only return the `chembl_id`, `molecule_properties`, and `molecule_structures` fields for the molecule with ChEMBL ID `CHEMBL25`. This reduces the amount of data returned and can speed up the query.

In [30]:
A = molecule.filter(chembl_id='CHEMBL25').only(['chembl_id' 'molecule_properties', 'molecule_structures'])
B = molecule.filter(chembl_id='CHEMBL25').only([ 'molecule_properties', 'molecule_structures'])

print(f'keys of A: {A[0].keys()}\nkeys of B: {B[0].keys()}')
print('Keys of A and B are the same: ', A[0].keys() == B[0].keys())
print('Values of A and B are the same: ', A[0] == B[0])
print(A == B)

keys of A: dict_keys(['molecule_properties', 'molecule_structures'])
keys of B: dict_keys(['molecule_properties', 'molecule_structures'])
Keys of A and B are the same:  True
Values of A and B are the same:  True
False


- As shown above, `A` and `B` are essentially the same in terms of their keys and values, though `A==B` still returned `False`. This is because they are indeed different queries. In fact, even `a==b` below returns `False`.

In [34]:
a = molecule.filter(chembl_id='CHEMBL25').only(['molecule_properties', 'molecule_structures'])
b = molecule.filter(chembl_id='CHEMBL25').only(['molecule_properties', 'molecule_structures'])
a == b

False

- Note that specified fields have to exist in the endpoint against which only is executed.
  - An API endpoint is a specific URL or address at which a service can be accessed by applications to perform operations such as fetching data or sending data. Think of it as a specific function or resource you can interact with through a URL. In simpler terms, an API endpoint is like a door through which you can request specific information from a database or service.
  - The molecule endpoint might be something like https://www.ebi.ac.uk/chembl/api/data/molecule. Each endpoint corresponds to a particular type of data or functionality provided by the API.
  - The example above used the method `molecule` to create a client. There are a lot of other methods one can use, which can be checked by the following code:

In [35]:
available_resources = [resource for resource in dir(new_client) if not resource.startswith('_')]
print(available_resources)

['activity', 'activity_supplementary_data_by_activity', 'assay', 'assay_class', 'atc_class', 'binding_site', 'biotherapeutic', 'cell_line', 'chembl_id_lookup', 'chembl_release', 'compound_record', 'compound_structural_alert', 'description', 'document', 'document_similarity', 'drug', 'drug_indication', 'drug_warning', 'go_slim', 'image', 'mechanism', 'metabolism', 'molecule', 'molecule_form', 'official', 'organism', 'protein_classification', 'similarity', 'source', 'substructure', 'target', 'target_component', 'target_relation', 'tissue', 'xref_source']


For example, one can create a client using `activity`:

In [37]:
activity = new_client.activity
activities = activity.filter(molecule_chembl_id='CHEMBL25').only(['activity_id', 'assay_chembl_id', 'standard_value', 'standard_units', 'standard_type'])
print(activities)

[{'activity_id': 42173, 'assay_chembl_id': 'CHEMBL702885', 'standard_type': "Log K'", 'standard_units': None, 'standard_value': '1.39', 'type': "Log K'", 'units': None, 'value': '1.39'}, {'activity_id': 81965, 'assay_chembl_id': 'CHEMBL771355', 'standard_type': 'Activity', 'standard_units': '%', 'standard_value': '100.0', 'type': 'Activity', 'units': '%', 'value': '100.0'}, {'activity_id': 81966, 'assay_chembl_id': 'CHEMBL780106', 'standard_type': 'ED25', 'standard_units': 'uM kg-1', 'standard_value': '402.0', 'type': 'ED25', 'units': 'uM kg-1', 'value': '402.0'}, {'activity_id': 88326, 'assay_chembl_id': 'CHEMBL762032', 'standard_type': 'IC50', 'standard_units': 'nM', 'standard_value': '34600.0', 'type': 'IC50', 'units': 'uM', 'value': '34.6'}, '...(remaining elements truncated)...']


or using `target`:

In [38]:
target = new_client.target
target_info = target.filter(target_chembl_id='CHEMBL2093871').only(['target_chembl_id', 'pref_name', 'organism', 'target_type', 'species_group_flag'])
print(target_info)

[{'organism': 'Rattus norvegicus', 'pref_name': 'Glutamate receptor ionotropic, AMPA', 'species_group_flag': False, 'target_chembl_id': 'CHEMBL2093871', 'target_type': 'PROTEIN COMPLEX GROUP'}]


In [15]:
A = {'one': 1, 'two': 2, 'three': 3}
B = {'three': 3, 'two': 2, 'one': 1}

# A == B is True
print(A == B)  # True

# A.values() == B.values() is False because the order of values is different
print(A.values() == B.values())  # False


True
False


### Limitations
- **Nested fields**: Specifying nested fields won't work as intended, as the `only` method ignores nested fields. The parent field will be returned instead. For instance `only(['molecule_proper__alogp'])` is treated as `only(['molecule_properties'])`.
- **Many-to-many relationship**: For many-to-many relationship, `only` won't optimize SQL joins. This means it won't reduce the complexity of retrieving related data in such relationships.  

### Benefits
By using `only`, you can significantly reduce the payload size and improve the performance of your API calls, esepcially when you only need a few fields from a potentially large dataset. This is particularly useful in scenarios where bandwidth and response time are critical. (Behind the scene, the API logic will also check if any SQL joins are necessary to return the specified field and exclude unnecessary joins with critically improves performance.)

### Complementary notes
#### What are SQL joins?
(Notes taken from the response by ChatGPT-4o.)
- SQL is a standard language used to communicate with databases. It allows you to create, read, update, and delete data in a database. In SQL, a join is a way to combine rows from two or more tables based on a related column between them. 
- Types of SQL Joins:
  - `INNER JOIN`: Returns only the rows that have matching values in both tables.
  - `LEFT JOIN` (or `LEFT OUTER JOIN`): Returns all rows from the left table, and the matched rows from the right table. If no match, NULL values are returned for columns from the right table.
  - `RIGHT JOIN` (or `RIGHT OUTER JOIN`): Returns all rows from the right table, and the matched rows from the left table. If no match, NULL values are returned for columns from the left table.
  - `FULL JOIN` (or `FULL OUTER JOIN`): Returns all rows when there is a match in either left or right table. Rows that do not have a match in either table are also included, with NULLs in the columns of the non-matching table.
- Example
  - Imagine you have two tables in a database.
    - Table 1: Molecules 
      | chembl_id | name | 
      | :--------: | :--------: |
      | CHEMBL25 | Aspirin     | 
      | CHEMBL26 | Paracetamol     | 
    - Table 2: Molecule Properties
      | chembl_id |   property_name  | property_value |
      |:---------:|:----------------:|:--------------:|
      |  CHEMBL25 |       logP       |       1.2      |
      |  CHEMBL25 | molecular_weight |     180.16     |
      |  CHEMBL26 |       logP       |       0.5      |
      | CHEMBL26  | molecular_weight | 151.17         |
  - Join example
    - You want to retrieve the molecule name and its logP value.
    - The common column between the two tables is `chembl_id`. 
  - SQL query
    ```sql
    SELECT molecule.name, molecule_properties.property_value AS logP
    FROM molecule
    INNER JOIN molecule_properties
    ON molecule.chembl_id = molecule_properties.chembl_id
    WHERE molecule_properties.property_name = 'logP';
    ```
  - Result
    |     name    | logP |
    |:-----------:|:----:|
    |   Aspirin   |  1.2 |
    | Paracetamol |  0.5 |
  - In this query:
    - `INNER JOIN` is used to combine rows from the molecule table and molecule_properties table where the `chembl_id` matches.
    - The `WHERE` clause filters the properties to only get the logP values.
  - When the API needs to fetch data that resides in multiple related tables, it may perform joins behind the scenes to combine the necessary data into a single response. By using the only method, you help the API understand which specific fields you need, allowing it to exclude unnecessary joins and optimize the query.

#### What are many-to-many relationships?
Many-to-many relationships are a common concept in database design where a record in one table can be associated with multiple records in another table, and vice versa. To manage many-to-many relationships, a third table, often called a "junction table" or "join table," is used to link the two related tables.
- Example scenario
  - Consider a database where we have three tables: "Compounds", "Assays" and "Targets". As shown below, the columns of the table create a many-to-many relationship between tables. Specifically
    - Compounds:
      | chembl_id | molecule_name |
      |:---------:|:-------------:|
      |  CHEMBL25 |    Aspirin    |
      |  CHEMBL26 |  Paracetamol  |
    - Assays:
      | assay_id |    assay_description   |
      |:--------:|:----------------------:|
      |    A1    | Inhibition of enzyme X |
      |    A2    |  Binding to receptor Y |
    - Targets:
      | target_id | target_name |
      |:---------:|:-----------:|
      |     T1    |   Enzyme X  |
      |     T2    |  Receptor Y |
  - Junction tables
    - The junction table for the compound-assay relationship
      | chembl_id | assay_id |
      |:---------:|:--------:|
      |  CHEMBL25 |    A1    |
      |  CHEMBL25 |    A2    |
      |  CHEMBL26 |    A1    |
    - The junction table for the assay-target relationship
      | assay_id | target_id |
      |:--------:|:---------:|
      |    A1    |     T1    |
      |    A2    |     T2    |
    - SQL joins for many-to-many relationships 
      To retrieve the list of compounds, their assays, and the associated targets, one can use SQL joins so that the following table is generated.
      | molecule_name |    assay_description   | target_name |
      |:-------------:|:----------------------:|:-----------:|
      |    Aspirin    | Inhibition of enzyme X |   Enzyme X  |
      |    Aspirin    |  Binding to receptor Y |  Receptor Y |
      |  Paracetamol  | Inhibition of enzyme X |   Enzyme X  |
